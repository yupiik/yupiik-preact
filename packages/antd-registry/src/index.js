import * as antdIcons from '@ant-design/icons';
import * as antd from 'antd';
import { simpleComponent } from '@yupiik/dynamic/src/component/registry';

const isComponent = it => {
    const c = it.charAt(0);
    return c === c.toUpperCase() && c !== '$' && c !== '_' && !it.endsWith('Props');
};

const antdComponent = nameOrComponent => {
    const impl = simpleComponent(nameOrComponent);
    /* can forward static props - not that helpful since we have 2 HoC wrapping the actual component
       but antd checks if the child is directly an "Item" sometimes so we can probably hack up a proxy there?
    Object
        .keys(nameOrComponent)
        .filter(it => it.startsWith('__ANT_'))
        .forEach(i => impl[i] = nameOrComponent[i]);
    */
    return impl;
}

const forward = obj => Object
    .keys(obj)
    .filter(it => isComponent(it))
    .reduce((a, i) => {
        const wrapper = obj[i];

        a[i] = antdComponent(wrapper);

        // sub-components (Header, Content, Footer, Item, ...)
        Object
            .keys(wrapper)
            .filter(it => isComponent(it))
            .forEach(it => {
                a[`${i}.${it}`] = antdComponent(wrapper[it]);
            });

        return a;
    }, {});

export const AntdRegistry = {
    ...forward(antd),
    ...forward(antdIcons),
};
