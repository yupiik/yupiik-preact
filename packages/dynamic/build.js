const builder = require('../../esbuild.config.base');

const copyUtilities = () => {
    const fs = require('fs');
    if (!fs.existsSync('dist/esbuild')) {
        fs.mkdirSync('dist/esbuild');
    }
    fs.copyFileSync('src/esbuild/shim.js', 'dist/esbuild/shim.js');
};

(async function build() {
    await builder.build();
    copyUtilities();
})();
