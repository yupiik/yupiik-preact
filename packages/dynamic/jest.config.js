module.exports = require('../../jest.config.base');
module.exports = require('../../jest.config.base');
