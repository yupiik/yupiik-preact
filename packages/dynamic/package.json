{
  "name": "@yupiik/dynamic",
  "description": "A simple preact based library to help configuring an UI.",
  "main": "dist/index.js",
  "mainSrc": "src/index.js",
  "author": "Yupiik <contact@yupiik.com>",
  "license": "ISC",
  "keywords": [
    "preact",
    "yupiik"
  ],
  "scripts": {
    "clean": "rm -Rf dist",
    "build": "node build.js",
    "publish": "npm publish --access public",
    "test": "jest",
    "test:watch": "npx jest --watchAll",
    "test:updateSnapshot": "npx jest --updateSnapshot",
    "test:debug": "node --inspect-brk ./node_modules/jest/bin/jest.js --runInBand"
  }
}
