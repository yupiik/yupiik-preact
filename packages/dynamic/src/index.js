import {
    Dynamic,
    ComponentRegistryContext
} from './component/Dynamic';
import {
    FromConfiguratonHoc,
} from './component/FromConfiguratonHoc';

import {
    HtmlRegistry,
} from './component/registry';

export {
    Dynamic,
    ComponentRegistryContext,
    FromConfiguratonHoc,
    HtmlRegistry,
};
