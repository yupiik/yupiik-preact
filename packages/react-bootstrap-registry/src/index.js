import { simpleComponent } from '@yupiik/dynamic/src/component/registry';
import * as reactBootstrap from 'react-bootstrap';

const isComponent = it =>
    it.charAt(0) === it.charAt(0).toUpperCase() &&
    !it.includes('$') &&
    !it.includes('_') &&
    !it.endsWith('Props');

const forward = obj => Object
    .keys(obj)
    .filter(it => isComponent(it))
    .reduce((a, i) => {
        const wrapper = obj[i];

        a[i] = simpleComponent(wrapper);

        // subcomponents
        Object
            .keys(wrapper)
            .filter(it => isComponent(it))
            .forEach(it => {
                a[`${i}.${it}`] = simpleComponent(wrapper[it]);
            });

        return a;
    }, {});

export const ReactBootstrapRegistry = forward(reactBootstrap);
