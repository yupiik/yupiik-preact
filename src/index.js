import {
    Dynamic,
    ComponentRegistryContext
} from './component/Dynamic';

export {
    Dynamic,
    ComponentRegistryContext,
};
