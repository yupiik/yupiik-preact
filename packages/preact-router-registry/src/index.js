import { h } from 'preact';
import { simpleComponent } from '@yupiik/dynamic';
import { Link, Router, route } from 'preact-router';

// avoids page reloading
const NavLink = ({ href, children, ...rest }) => (
    <a href={href || ''} {...rest} onClick={e => {
        e.preventDefault();
        route(href || '');
    }}>
        {children || ''}
    </a>);

export const PreactRouterRegistry = {
    Router: simpleComponent(Router),
    Link: simpleComponent(Link),
    NavLink: simpleComponent(NavLink),
};
