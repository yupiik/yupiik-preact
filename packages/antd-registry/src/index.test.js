import { h } from 'preact';
import { render } from '@testing-library/preact';
import { expect } from 'expect';
import { FromConfiguratonHoc, HtmlRegistry } from '../../dynamic/src';
import { AntdRegistry } from './index';

test('Should render ant components', () => {
    // statc part (JSON friendly)
    const header = {
        options: {
            name: 'Layout.Header',
            options: {
                configuration: {
                    children: [
                        {
                            name: 'div',
                            options: {
                                configuration: {
                                    className: 'logo',
                                },
                            },
                        },
                        {
                            name: 'Menu',
                            options: {
                                configuration: {
                                    theme: 'dark',
                                    mode: 'horizontal',
                                    items: [
                                        {
                                            label: 'sub menu',
                                            key: 'submenu',
                                        },
                                    ],
                                },
                            },
                        },
                    ],
                },
            },
        },
    };
    const routes = [
        {
            name: 'home',
            breadcrumbName: 'Home',
            path: '/',
            childRoutes: [
                {
                    name: 'apps',
                    breadcrumbName: 'Application List',
                    path: 'apps',
                    childRoutes: [
                        {
                            name: 'app',
                            breadcrumbName: 'Application:id',
                            path: ':id',
                            childRoutes: [
                                {
                                    name: 'detail',
                                    breadcrumbName: 'Detail',
                                    path: 'detail',
                                },
                            ],
                        },
                    ],
                },
            ],
        },
        {
            name: 'apps',
            breadcrumbName: 'Application List',
            path: 'apps',
            childRoutes: [
                {
                    name: 'app',
                    breadcrumbName: 'Application:id',
                    path: ':id',
                    childRoutes: [
                        {
                            name: 'detail',
                            breadcrumbName: 'Detail',
                            path: 'detail',
                        },
                    ],
                },
            ],
        },
    ];
    const breadcrumb = {
        options: {
            name: 'Breadcrumb',
            options: {
                configuration: {
                    style: { margin: '16px 0' },
                    routes,
                },
            },
        },
    };
    const content = {
        options: {
            name: 'Layout.Content',
            options: {
                configuration: {
                    style: { padding: '0 50px' },
                    children: [
                        breadcrumb,
                    ],
                },
            },
        },
    };
    const opts = {
        name: 'Layout',
        options: {
            configuration: {
                className: 'layout',
                children: [
                    header,
                    content,
                ],
            },
        },
    };

    // rendering
    const ui = (
        <FromConfiguratonHoc
            registry={{
                ...HtmlRegistry, // for <div />
                ...AntdRegistry,
            }}
            options={opts}
        />
    );
    const { container: { children } } = render(ui);

    // test
    expect(children).toMatchSnapshot();
});
