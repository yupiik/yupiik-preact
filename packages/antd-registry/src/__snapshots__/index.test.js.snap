// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ant components 1`] = `
HTMLCollection [
  <div
    class="ant-layout layout css-dev-only-do-not-override-nllxry"
  >
    <header
      class="ant-layout-header css-dev-only-do-not-override-nllxry"
    >
      <div
        class="logo"
      />
      <ul
        class="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-dark css-dev-only-do-not-override-nllxry"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-submenu"
          role="menuitem"
          style="opacity: 1; order: 0;"
          tabindex="-1"
        >
          <span
            class="ant-menu-title-content"
          >
            sub menu
          </span>
        </li>
        <li
          aria-hidden="true"
          class="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal ant-menu-submenu-disabled"
          role="none"
          style="opacity: 0; height: 0px; overflow-y: hidden; order: 9007199254740991; pointer-events: none; position: absolute;"
        >
          <div
            aria-controls="rc-menu-uuid-test-rc-menu-more-popup"
            aria-disabled="true"
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            data-menu-id="rc-menu-uuid-test-rc-menu-more"
            role="menuitem"
          >
            <span
              aria-label="ellipsis"
              class="anticon anticon-ellipsis"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="ellipsis"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                />
              </svg>
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </header>
    <main
      class="ant-layout-content css-dev-only-do-not-override-nllxry"
      style="padding: 0px 50px;"
    >
      <nav
        class="ant-breadcrumb css-dev-only-do-not-override-nllxry"
        style="margin: 16px 0px;"
      >
        <ol>
          <li>
            <a
              class="ant-breadcrumb-link"
              href="#/"
            >
              Home
            </a>
          </li>
          <li
            aria-hidden="true"
            class="ant-breadcrumb-separator"
          >
            /
          </li>
          <li>
            <a
              class="ant-breadcrumb-link"
              href="#//apps"
            >
              Application List
            </a>
          </li>
          
        </ol>
      </nav>
    </main>
  </div>,
]
`;
